### PR TITLE
feat: add configuration option to set max buffer size

### PR DESCRIPTION
**Pull request by imabhichow**
*Issue #, if available:*

*Description of changes:*

- Add configuration option to set max buffer size. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: src/main/java/software/amazon/encryption/s3/internal/BufferedCipherPublisher.java [hunk 1]
```diff
@@ -13,30 +12,23 @@ public class BufferedCipherPublisher implements SdkPublisher<ByteBuffer> {
...
         this.iv = iv;
+        this.bufferSize = bufferSize;
     }
 
     @Override
     public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
         // Wrap the (customer) subscriber in a CipherSubscriber, then subscribe it
         // to the wrapped (ciphertext) publisher
-        Subscriber<? super ByteBuffer> wrappedSubscriber = RangedGetUtils.adjustToDesiredRange(subscriber, range,
-                contentRange, cipherTagLengthBits);
```
Comment (imabhichow) [inline src/main/java/software/amazon/encryption/s3/internal/BufferedCipherPublisher.java:39]: Ranged Get is never called during buffer mode.

File: src/main/java/software/amazon/encryption/s3/internal/BufferedCipherSubscriber.java [hunk 2]
```diff
@@ -40,21 +36,22 @@ public class BufferedCipherSubscriber implements Subscriber<ByteBuffer> {
...
+    BufferedCipherSubscriber(Subscriber<? super ByteBuffer> wrappedSubscriber, Long contentLength, CryptographicMaterials materials, byte[] iv, long bufferSize) {
         this.wrappedSubscriber = wrappedSubscriber;
         if (contentLength == null) {
             throw new S3EncryptionClientException("contentLength cannot be null in buffered mode. To enable unbounded " +
                     "streaming, reconfigure the S3 Encryption Client with Delayed Authentication mode enabled.");
         }
-        if (contentLength > BUFFERED_MAX_CONTENT_LENGTH_BYTES) {
+        long bufferSizeInBytes = 1024 * 1024 * bufferSize;
+        if (contentLength > bufferSizeInBytes) {
             throw new S3EncryptionClientException(String.format("The object you are attempting to decrypt exceeds the maximum content " +
```
Comment (justplaz) [inline src/main/java/software/amazon/encryption/s3/internal/BufferedCipherSubscriber.java:47]: nit: we can update this exception text to reflect the configurable nature of the buffer size e.g.:

```
"The object you are attempting to decrypt exceeds the maximum buffer size: " + bufferSize +
"for the default (buffered) mode. Either increase your buffer size when configuring your client, " + 
"or enable Delayed Authentication mode to disable buffered decryption."
```

File: src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java [hunk 3]
```diff
@@ -40,8 +40,8 @@
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.api.Assertions.fail;
-import static software.amazon.encryption.s3.utils.S3EncryptionClientTestResources.appendTestSuffix;
-import static software.amazon.encryption.s3.utils.S3EncryptionClientTestResources.deleteObject;
+import static software.amazon.encryption.s3.utils.S3EncryptionClientTestResources.*;
+import static software.amazon.encryption.s3.utils.S3EncryptionClientTestResources.KMS_KEY_ID;
```
Comment (justplaz) [inline src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java:44]: nit - do we actually need this, if we have a `*` import from the class?

File: src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java [hunk 4]
```diff
@@ -203,6 +203,69 @@ public void ordinaryInputStreamV3DecryptCbc() throws IOException {
...
+                .cryptoProvider(provider)
+                .maxBufferSize(32)
+                .build();
+
+        // V3 Client with default buffer size (i.e. 64MiB)
+        S3Client v3ClientWithDelayedAuth = S3EncryptionClient.builder()
+                .aesKey(AES_KEY)
+                .cryptoProvider(provider)
+                .maxBufferSize(32)
+                .enableDelayedAuthenticationMode(true)
```
Comment (justplaz) [inline src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java:237]: We should validate against this case. When `enableDelayedAuthenticationMode` is set to `true` we don't buffer anything, so it shouldn't be possible to configure `maxBufferSize`.

**Review by justplaz — CHANGES_REQUESTED** [review]

File: src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java [hunk 5]
```diff
@@ -245,6 +248,11 @@ public void close() {
     // This is very similar to the S3EncryptionClient builder
     // Make sure to keep both clients in mind when adding new builder options
     public static class Builder {
+        private static final int MIN_ALLOWED_BUFFER_SIZE_MiB = 32;
```
[the remark this answers is not in the document]
Comment (imabhichow) [reply src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java:251]: agreed.

File: src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java (hunk 4 above)
Comment (imabhichow) [reply src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java:237]: agreed.

Comment (imabhichow) [reply src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java:237]: Just to confirm, we shouldn't throw an exception even though both `maxBufferSize` and `enableDelayedAuthenticationMode` set to true. But, if `enableDelayedAuthenticationMode` is enabled, this mode always takes priority over Buffered Mode.

Comment (justplaz) [reply src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java]: If `enableDelayedAuthenticationMode` is set to `true` AND `maxBufferSize` is set, we should throw an exception. It shouldn't be possible to configure the buffer size, when there is no buffer to configure.

File: src/main/java/software/amazon/encryption/s3/S3EncryptionClient.java [hunk 6]
```diff
@@ -676,6 +683,22 @@ public Builder enableMultipartPutObject(boolean _enableMultipartPutObject) {
...
         }
 
+        /**
+         * Sets the buffer size for safe authentication used when delayed authentication mode is disabled.
+         * If buffer size is not given during client configuration, default buffer size is set to 64MiB.
+         * @param bufferSize the desired buffer size in Bytes.
+         * @return Returns a reference to this object so that method calls can be chained together.
+         * @throws S3EncryptionClientException if the specified buffer size is outside the allowed bounds
+         */
+        public Builder maxBufferSize(long bufferSize) throws IllegalArgumentException {
```
Comment (justplaz) [inline src/main/java/software/amazon/encryption/s3/S3EncryptionClient.java:693]: I think we should rename this to `setBufferSize`, it makes more sense in the context of the client builder

File: src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java [hunk 7]
```diff
@@ -434,6 +441,22 @@ public Builder enableMultipartPutObject(boolean _enableMultipartPutObject) {
...
         }
 
+        /**
+         * Sets the buffer size for safe authentication used when delayed authentication mode is disabled.
+         * If buffer size is not given during client configuration, default buffer size is set to 64MiB.
+         * @param bufferSize the desired buffer size in Bytes.
+         * @return Returns a reference to this object so that method calls can be chained together.
+         * @throws S3EncryptionClientException if the specified buffer size is outside the allowed bounds
+         */
+        public Builder maxBufferSize(long bufferSize) throws IllegalArgumentException {
```
Comment (justplaz) [inline src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java:451]: ditto - rename (see other comment)

Comment (justplaz) [reply src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java:451]: does this ever throw `IllegalArgumentException`?

File: src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java [hunk 8]
```diff
@@ -257,6 +263,7 @@ public static class Builder {
         private boolean _enableMultipartPutObject = false;
         private Provider _cryptoProvider = null;
         private SecureRandom _secureRandom = new SecureRandom();
+        private long _bufferSize = 0;
```
Comment (justplaz) [inline src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java:266]: nit: I prefer `-1` as the value for "not used" (see how `contentLength` is handled elsewhere)

File: src/main/java/software/amazon/encryption/s3/S3EncryptionClient.java (hunk 6 above)
Comment (justplaz) [reply src/main/java/software/amazon/encryption/s3/S3EncryptionClient.java:693]: does this ever throw `IllegalArgumentException`?

File: src/main/java/software/amazon/encryption/s3/internal/BufferedCipherPublisher.java [hunk 9]
```diff
@@ -13,30 +12,23 @@ public class BufferedCipherPublisher implements SdkPublisher<ByteBuffer> {
...
         this.materials = materials;
         this.iv = iv;
+        this.bufferSize = bufferSize;
     }
 
     @Override
     public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
         // Wrap the (customer) subscriber in a CipherSubscriber, then subscribe it
         // to the wrapped (ciphertext) publisher
-        Subscriber<? super ByteBuffer> wrappedSubscriber = RangedGetUtils.adjustToDesiredRange(subscriber, range,
```
Comment (justplaz) [inline src/main/java/software/amazon/encryption/s3/internal/BufferedCipherPublisher.java:38]: nice!

File: src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java [hunk 10]
```diff
@@ -203,6 +204,78 @@ public void ordinaryInputStreamV3DecryptCbc() throws IOException {
...
+                .bucket(BUCKET)
+                .key(objectKey));
+
+
+        assertTrue(IOUtils.contentEquals(new BoundedInputStream(fileSizeExceedingDefaultLimit), response));
+        response.close();
+
+        // Cleanup
+        deleteObject(BUCKET, objectKey, v3ClientWithBuffer32MiB);
+        v3ClientWithBuffer32MiB.close();
```
Comment (justplaz) [inline src/test/java/software/amazon/encryption/s3/S3EncryptionClientStreamTest.java:276]: nit: close both clients

File: src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java (hunk 7 above)
Comment (justplaz) [reply src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java:451]: Async class isn't tested; let's at least add some basic tests for the async client.

**Review by justplaz — CHANGES_REQUESTED** [review]
overall, looks good, just some nits and naming comments 
also - can we delete `src/main/java/software/amazon/encryption/s3/internal/BufferedAesGcmContentStrategy.java` - seems it's still present but not used anywhere, we might as well remove it - thanks!

File: src/main/java/software/amazon/encryption/s3/S3EncryptionClient.java (hunk 6 above)
Comment (imabhichow) [reply src/main/java/software/amazon/encryption/s3/S3EncryptionClient.java:693]: It's a nit, will remove `IllegalArgumentException` and clean up the code.

File: src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java (hunk 8 above)
Comment (imabhichow) [reply src/main/java/software/amazon/encryption/s3/S3AsyncEncryptionClient.java:266]: ack!

**Review by justplaz — APPROVED** [review]
LGTM!

**Pull request merged**